File: packages/ios-im-ui/Sources/FlareIMUI/Models/MessageGrouping.swift
import Foundation

public enum MessageGroupPosition: String, Sendable { case single, first, middle, last }
public enum MessageAvatarPlacement: String, Sendable { case leading, trailing }

public struct MessageRowPresentation: Sendable {
    public let showAvatar: Bool
    public let reserveAvatarSpace: Bool
    public let showSenderName: Bool
    public let avatarPlacement: MessageAvatarPlacement

    public init(
        showAvatar: Bool = false,
        reserveAvatarSpace: Bool = false,
        showSenderName: Bool = false,
        avatarPlacement: MessageAvatarPlacement = .leading
    ) {
        self.showAvatar = showAvatar; self.reserveAvatarSpace = reserveAvatarSpace
        self.showSenderName = showSenderName; self.avatarPlacement = avatarPlacement
    }
}

public func messagesShareGroup(
    _ previous: FlareMessageData?,
    _ current: FlareMessageData?,
    timeGapMs: Int64 = 300_000
) -> Bool {
    // Notices (system lines, recalled messages) stand alone and end a sender's run.
    guard let previous, let current, !previous.isSystem, !current.isSystem, !previous.isRecalled, !current.isRecalled else { return false }
    guard previous.senderId == current.senderId else { return false }
    guard previous.sentAtMs > 0, current.sentAtMs > 0 else { return true }
    return abs(current.sentAtMs - previous.sentAtMs) <= timeGapMs
}

public func messageGroupPosition(
    _ messages: [FlareMessageData],
    index: Int,
    timeGapMs: Int64 = 300_000
) -> MessageGroupPosition {
    let current = messages[index]
    if current.isSystem || current.isRecalled { return .single }
    let previous = index > 0 ? messages[index - 1] : nil
    let next = index + 1 < messages.count ? messages[index + 1] : nil
    let joinsPrevious = messagesShareGroup(previous, current, timeGapMs: timeGapMs)
    let joinsNext = messagesShareGroup(current, next, timeGapMs: timeGapMs)
    if !joinsPrevious && !joinsNext { return .single }
    if !joinsPrevious { return .first }
    if !joinsNext { return .last }
    return .middle
}

public func messageRowPresentation(
    message: FlareMessageData,
    position: MessageGroupPosition,
    currentUserId: String,
    groupConversation: Bool,
    showIncomingAvatar: Bool = true,
    showSelfAvatar: Bool = false,
    showGroupSenderName: Bool = true
) -> MessageRowPresentation {
    let own = message.senderId == currentUserId
    let edge = position == .single || position == .first
    return MessageRowPresentation(
        showAvatar: (own ? showSelfAvatar : showIncomingAvatar) && edge,
        reserveAvatarSpace: own ? showSelfAvatar : showIncomingAvatar,
        showSenderName: !own && groupConversation && showGroupSenderName && edge,
        avatarPlacement: own ? .trailing : .leading
    )
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/MessageGrouping.kt
package com.flare.im.ui

enum class MessageGroupPosition { Single, First, Middle, Last }

enum class MessageAvatarPlacement { Leading, Trailing }

data class MessageRowPresentation(
    val showAvatar: Boolean = false,
    val reserveAvatarSpace: Boolean = false,
    val showSenderName: Boolean = false,
    val avatarPlacement: MessageAvatarPlacement = MessageAvatarPlacement.Leading,
)

fun messagesShareGroup(previous: FlareMessageData?, current: FlareMessageData?, timeGapMs: Long = 300_000): Boolean {
    // Notices (system lines, recalled messages) stand alone and end a sender's run.
    if (previous == null || current == null || previous.isSystem || current.isSystem || previous.isRecalled || current.isRecalled) return false
    if (previous.senderId != current.senderId) return false
    if (previous.sentAtMs <= 0 || current.sentAtMs <= 0) return true
    return kotlin.math.abs(current.sentAtMs - previous.sentAtMs) <= timeGapMs
}

fun messageGroupPosition(messages: List<FlareMessageData>, index: Int, timeGapMs: Long = 300_000): MessageGroupPosition {
    val current = messages[index]
    if (current.isSystem || current.isRecalled) return MessageGroupPosition.Single
    val joinsPrevious = messagesShareGroup(messages.getOrNull(index - 1), current, timeGapMs)
    val joinsNext = messagesShareGroup(current, messages.getOrNull(index + 1), timeGapMs)
    return when {
        !joinsPrevious && !joinsNext -> MessageGroupPosition.Single
        !joinsPrevious -> MessageGroupPosition.First
        !joinsNext -> MessageGroupPosition.Last
        else -> MessageGroupPosition.Middle
    }
}

fun messageRowPresentation(
    message: FlareMessageData,
    position: MessageGroupPosition,
    currentUserId: String,
    groupConversation: Boolean,
    showIncomingAvatar: Boolean = true,
    showSelfAvatar: Boolean = false,
    showGroupSenderName: Boolean = true,
): MessageRowPresentation {
    val self = message.senderId == currentUserId
    val edge = position == MessageGroupPosition.Single || position == MessageGroupPosition.First
    return MessageRowPresentation(
        showAvatar = (if (self) showSelfAvatar else showIncomingAvatar) && edge,
        reserveAvatarSpace = if (self) showSelfAvatar else showIncomingAvatar,
        showSenderName = !self && groupConversation && showGroupSenderName && edge,
        avatarPlacement = if (self) MessageAvatarPlacement.Trailing else MessageAvatarPlacement.Leading,
    )
}
